# Milestone 1
Group 4: Mengjun Chen, Qurat-ul-Ain Azim, Ashwin Babu, Mohammad Reza Nabizadeh
## 1. Team-work contract
- Link to the team work contract:
https://docs.google.com/document/d/1FKmLnss40QeTBvXecWfMuavG4hm0lxKGkV9BtIFVIS8/edit
## 2. Creating a repository and project structure
- Link to the repository
https://github.com/mrnabiz/dsci_525_group_4

## 3. Downloading the data

In [1]:
import re
import os
import glob
import zipfile
import requests
from urllib.request import urlretrieve
import json
import pandas as pd

In [2]:
%cd ~/Desktop/MDS/525/dsci_525_group_4

/Users/mrnabiz/Desktop/MDS/525/dsci_525_group_4


In [3]:
# Necessary metadata
article_id = 14096681  # this is the unique identifier of the article on figshare
url = f"https://api.figshare.com/v2/articles/{article_id}"
headers = {"Content-Type": "application/json"}
output_directory = "figshare/"

In [4]:
response = requests.request("GET", url, headers=headers)
data = json.loads(response.text)  # this contains all the articles data, feel free to check it out
files = data["files"]             # this is just the data about the files, which is what we want
files

[{'id': 26579150,
  'name': 'daily_rainfall_2014.png',
  'size': 58863,
  'is_link_only': False,
  'download_url': 'https://ndownloader.figshare.com/files/26579150',
  'supplied_md5': 'fd32a2ffde300a31f8d63b1825d47e5e',
  'computed_md5': 'fd32a2ffde300a31f8d63b1825d47e5e'},
 {'id': 26579171,
  'name': 'environment.yml',
  'size': 192,
  'is_link_only': False,
  'download_url': 'https://ndownloader.figshare.com/files/26579171',
  'supplied_md5': '060b2020017eed93a1ee7dd8c65b2f34',
  'computed_md5': '060b2020017eed93a1ee7dd8c65b2f34'},
 {'id': 26586554,
  'name': 'README.md',
  'size': 5422,
  'is_link_only': False,
  'download_url': 'https://ndownloader.figshare.com/files/26586554',
  'supplied_md5': '61858c6cc0e6a6d6663a7e4c75bbd88c',
  'computed_md5': '61858c6cc0e6a6d6663a7e4c75bbd88c'},
 {'id': 26766812,
  'name': 'data.zip',
  'size': 814041183,
  'is_link_only': False,
  'download_url': 'https://ndownloader.figshare.com/files/26766812',
  'supplied_md5': 'b517383f76e77bd03755a63a8f

In [5]:
%%time
files_to_dl = ["data.zip"]  # feel free to add other files here
for file in files:
    if file["name"] in files_to_dl:
        os.makedirs(output_directory, exist_ok=True)
        urlretrieve(file["download_url"], output_directory + file["name"])

CPU times: user 2.77 s, sys: 2.72 s, total: 5.49 s
Wall time: 57.3 s


In [6]:
%%time
with zipfile.ZipFile(os.path.join(output_directory, "data.zip"), 'r') as f:
    f.extractall(output_directory)

CPU times: user 12.5 s, sys: 1.55 s, total: 14 s
Wall time: 15.1 s


In [10]:
%ls -ltr figshare/

total 12426752
drwxr-xr-x  30 mrnabiz  staff        960 27 Mar 15:10 __MACOSX/
-rw-r--r--   1 mrnabiz  staff  814041183 27 Mar 21:27 data.zip
-rw-r--r--   1 mrnabiz  staff   95376895 27 Mar 21:27 MPI-ESM-1-2-HAM_daily_rainfall_NSW.csv
-rw-r--r--   1 mrnabiz  staff   94960113 27 Mar 21:27 AWI-ESM-1-1-LR_daily_rainfall_NSW.csv
-rw-r--r--   1 mrnabiz  staff   82474546 27 Mar 21:27 NorESM2-LM_daily_rainfall_NSW.csv
-rw-r--r--@  1 mrnabiz  staff  127613760 27 Mar 21:27 ACCESS-CM2_daily_rainfall_NSW.csv
-rw-r--r--   1 mrnabiz  staff  232118894 27 Mar 21:27 FGOALS-f3-L_daily_rainfall_NSW.csv
-rw-r--r--   1 mrnabiz  staff  330360682 27 Mar 21:27 CMCC-CM2-HR4_daily_rainfall_NSW.csv
-rw-r--r--   1 mrnabiz  staff  254009247 27 Mar 21:27 MRI-ESM2-0_daily_rainfall_NSW.csv
-rw-r--r--   1 mrnabiz  staff  235661418 27 Mar 21:27 GFDL-CM4_daily_rainfall_NSW.csv
-rw-r--r--   1 mrnabiz  staff  294260911 27 Mar 21:27 BCC-CSM2-MR_daily_rainfall_NSW.csv
-rw-r--r--   1 mrnabiz  staff  295768615 27 Mar 21:27 E

## 4. Combining data CSVs

In [11]:
%%time
## here we are using a normal python way for merging the data 
import pandas as pd
use_cols = ["time", "lat_min", "lat_max", "lon_min", "lon_max", "rain (mm/day)"]
files = glob.glob('figshare/*.csv')
df = pd.concat((pd.read_csv(file, index_col=0, usecols=use_cols)
                .assign(model=re.findall("/([^_]*)", file)[0])
                for file in files)
              )
df.to_csv("figshare/combined_data.csv")

CPU times: user 5min 8s, sys: 13 s, total: 5min 21s
Wall time: 5min 24s


### Results Comparison:
- **Mengjun Chen:**

- **Qurat-ul-Ain Azim:**

- **Ashwin Babu:**

- **Mohammad Reza Nabizadeh:**
The task to combine all of the CSVs, add a new column to the combined file, and save it as a new file took me 5 mins and 24 seconds.



| Team Member | Operating System | RAM | Processor | Is SSD | Time taken |
|:-----------:|:----------------:|:---:|:---------:|:------:|:----------:|
| Mengjun Chen    | MacOS                 |     |           |        |            |
| Qurat-ul-Ain Azim    |                  |     |           |        |            |
| Ashwin Babu    |                  |     |           |        |            |
| Mohammad Reza Nabizadeh | MacOS | 16GB | Apple M2 | YES | 5 Mins and 24 Seconds|

## 5. Load the combined CSV to memory and perform a simple EDA
### Changing `dtype`of data

In [23]:
use_cols = ["time", "lat_min", "lat_max", "lon_min", "lon_max", "rain (mm/day)", "model"]
df = pd.read_csv("figshare/combined_data.csv",usecols=use_cols)

In [24]:
%%time
print(df["model"].value_counts())

MPI-ESM1-2-HR       5154240
CMCC-CM2-HR4        3541230
CMCC-ESM2           3541230
CMCC-CM2-SR5        3541230
NorESM2-MM          3541230
TaiESM1             3541230
SAM0-UNICON         3541153
GFDL-ESM4           3219300
FGOALS-f3-L         3219300
GFDL-CM4            3219300
MRI-ESM2-0          3037320
EC-Earth3-Veg-LR    3037320
BCC-CSM2-MR         3035340
MIROC6              2070900
ACCESS-CM2          1932840
ACCESS-ESM1-5       1610700
INM-CM4-8           1609650
INM-CM5-0           1609650
FGOALS-g3           1287720
KIOST-ESM           1287720
AWI-ESM-1-1-LR       966420
MPI-ESM1-2-LR        966420
NESM3                966420
MPI-ESM-1-2-HAM      966420
NorESM2-LM           919800
BCC-ESM1             551880
CanESM5              551880
Name: model, dtype: int64
CPU times: user 3.1 s, sys: 51.6 ms, total: 3.15 s
Wall time: 3.14 s


In [25]:
%%time
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62467843 entries, 0 to 62467842
Data columns (total 7 columns):
 #   Column         Dtype  
---  ------         -----  
 0   time           object 
 1   lat_min        float64
 2   lat_max        float64
 3   lon_min        float64
 4   lon_max        float64
 5   rain (mm/day)  float64
 6   model          object 
dtypes: float64(5), object(2)
memory usage: 3.3+ GB
CPU times: user 2.97 ms, sys: 1.23 ms, total: 4.2 ms
Wall time: 4.35 ms


In [26]:
%%time
df.describe()

CPU times: user 11.3 s, sys: 3.36 s, total: 14.7 s
Wall time: 15 s


lat_min       lat_max       lon_min       lon_max  rain (mm/day)
count  5.924854e+07  6.246784e+07  5.924854e+07  6.246784e+07   5.924854e+07
mean  -3.310482e+01 -3.197757e+01  1.469059e+02  1.482150e+02   1.901170e+00
std    1.963549e+00  1.992067e+00  3.793784e+00  3.809994e+00   5.585735e+00
min   -3.646739e+01 -3.600000e+01  1.406250e+02  1.412500e+02  -3.807373e-12
25%   -3.486911e+01 -3.366221e+01  1.434375e+02  1.450000e+02   3.838413e-06
50%   -3.300000e+01 -3.204188e+01  1.468750e+02  1.481250e+02   6.154947e-02
75%   -3.140170e+01 -3.015707e+01  1.501875e+02  1.513125e+02   1.020918e+00
max   -2.990000e+01 -2.790606e+01  1.537500e+02  1.556250e+02   4.329395e+02

In [27]:
print(f"Memory usage with float64: {df[['lat_min', 'lat_max', 'lon_min' ,'lon_max', 'rain (mm/day)', 'model']].memory_usage().sum() / 1e6:.2f} MB")
print(f"Memory usage with float32: {df[['lat_min', 'lat_max', 'lon_min' ,'lon_max', 'rain (mm/day)', 'model']].astype('float32', errors='ignore').memory_usage().sum() / 1e6:.2f} MB")

Memory usage with float64: 2998.46 MB
Memory usage with float32: 1749.10 MB


In [28]:
df[['lat_min', 'lat_max', 'lon_min' ,'lon_max', 'rain (mm/day)', 'model']] = \
df[['lat_min', 'lat_max', 'lon_min' ,'lon_max', 'rain (mm/day)', 'model']].astype('float32', errors='ignore')

In [32]:
%%time
print(df["model"].value_counts())

MPI-ESM1-2-HR       5154240
CMCC-CM2-HR4        3541230
CMCC-ESM2           3541230
CMCC-CM2-SR5        3541230
NorESM2-MM          3541230
TaiESM1             3541230
SAM0-UNICON         3541153
GFDL-ESM4           3219300
FGOALS-f3-L         3219300
GFDL-CM4            3219300
MRI-ESM2-0          3037320
EC-Earth3-Veg-LR    3037320
BCC-CSM2-MR         3035340
MIROC6              2070900
ACCESS-CM2          1932840
ACCESS-ESM1-5       1610700
INM-CM4-8           1609650
INM-CM5-0           1609650
FGOALS-g3           1287720
KIOST-ESM           1287720
AWI-ESM-1-1-LR       966420
MPI-ESM1-2-LR        966420
NESM3                966420
MPI-ESM-1-2-HAM      966420
NorESM2-LM           919800
BCC-ESM1             551880
CanESM5              551880
Name: model, dtype: int64
CPU times: user 3.08 s, sys: 22 ms, total: 3.1 s
Wall time: 3.09 s


In [33]:
%%time
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62467843 entries, 0 to 62467842
Data columns (total 7 columns):
 #   Column         Dtype  
---  ------         -----  
 0   time           object 
 1   lat_min        float32
 2   lat_max        float32
 3   lon_min        float32
 4   lon_max        float32
 5   rain (mm/day)  float32
 6   model          object 
dtypes: float32(5), object(2)
memory usage: 2.1+ GB
CPU times: user 6.43 ms, sys: 585 µs, total: 7.01 ms
Wall time: 6.81 ms


In [34]:
%%time
df.describe()

CPU times: user 10.5 s, sys: 1.31 s, total: 11.8 s
Wall time: 11.8 s


lat_min       lat_max       lon_min       lon_max  rain (mm/day)
count  5.924854e+07  6.246784e+07  5.924854e+07  6.246784e+07   5.924854e+07
mean  -3.310497e+01 -3.197765e+01  1.469057e+02  1.482150e+02   1.901173e+00
std    1.963549e+00  1.992067e+00  3.793784e+00  3.809994e+00   5.585735e+00
min   -3.646739e+01 -3.600000e+01  1.406250e+02  1.412500e+02  -3.807373e-12
25%   -3.486911e+01 -3.366221e+01  1.434375e+02  1.450000e+02   3.838413e-06
50%   -3.300000e+01 -3.204189e+01  1.468750e+02  1.481250e+02   6.154947e-02
75%   -3.140170e+01 -3.015707e+01  1.501875e+02  1.513125e+02   1.020918e+00
max   -2.990000e+01 -2.790606e+01  1.537500e+02  1.556250e+02   4.329395e+02

### Results Comparison:
- **Mengjun Chen:**

- **Qurat-ul-Ain Azim:**

- **Ashwin Babu:**

- **Mohammad Reza Nabizadeh:**
By changing the type the float values from float 64 to float 32:
  - The `value_counts` time dropped from 3.14 s to 3.09 s.
  - The `info` time dropped from 4.35 ms to 6.81 ms.
  - The `describe` time dropped from 15 s to 11.8 s.
  
### Load just columns that we want

In [36]:
use_cols = ["time", "lat_min", "lon_min", "rain (mm/day)", "model"]
df = pd.read_csv("figshare/combined_data.csv",usecols=use_cols)

In [40]:
%%time
print(df["model"].value_counts())

MPI-ESM1-2-HR       5154240
CMCC-CM2-HR4        3541230
CMCC-ESM2           3541230
CMCC-CM2-SR5        3541230
NorESM2-MM          3541230
TaiESM1             3541230
SAM0-UNICON         3541153
GFDL-ESM4           3219300
FGOALS-f3-L         3219300
GFDL-CM4            3219300
MRI-ESM2-0          3037320
EC-Earth3-Veg-LR    3037320
BCC-CSM2-MR         3035340
MIROC6              2070900
ACCESS-CM2          1932840
ACCESS-ESM1-5       1610700
INM-CM4-8           1609650
INM-CM5-0           1609650
FGOALS-g3           1287720
KIOST-ESM           1287720
AWI-ESM-1-1-LR       966420
MPI-ESM1-2-LR        966420
NESM3                966420
MPI-ESM-1-2-HAM      966420
NorESM2-LM           919800
BCC-ESM1             551880
CanESM5              551880
Name: model, dtype: int64
CPU times: user 3.08 s, sys: 19.5 ms, total: 3.1 s
Wall time: 3.09 s


In [41]:
%%time
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62467843 entries, 0 to 62467842
Data columns (total 5 columns):
 #   Column         Dtype  
---  ------         -----  
 0   time           object 
 1   lat_min        float64
 2   lon_min        float64
 3   rain (mm/day)  float64
 4   model          object 
dtypes: float64(3), object(2)
memory usage: 2.3+ GB
CPU times: user 5.69 ms, sys: 395 µs, total: 6.09 ms
Wall time: 5.87 ms


In [42]:
%%time
df.describe()

CPU times: user 7.3 s, sys: 1.56 s, total: 8.86 s
Wall time: 8.85 s


lat_min       lon_min  rain (mm/day)
count  5.924854e+07  5.924854e+07   5.924854e+07
mean  -3.310482e+01  1.469059e+02   1.901170e+00
std    1.963549e+00  3.793784e+00   5.585735e+00
min   -3.646739e+01  1.406250e+02  -3.807373e-12
25%   -3.486911e+01  1.434375e+02   3.838413e-06
50%   -3.300000e+01  1.468750e+02   6.154947e-02
75%   -3.140170e+01  1.501875e+02   1.020918e+00
max   -2.990000e+01  1.537500e+02   4.329395e+02

### Results Comparison:
- **Mengjun Chen:**

- **Qurat-ul-Ain Azim:**

- **Ashwin Babu:**

- **Mohammad Reza Nabizadeh:**
By just selecting the columns of the data we need:
  - The `value_counts` time dropped from 3.14 s to 3.09 s.
  - The `info` time dropped from 4.35 ms to 5.78 ms.
  - The `describe` time dropped from 15 s to 8.85 s.